flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java

```diff
@@ -273,7 +274,15 @@ public static SdkAsyncHttpClient createAsyncHttpClient(
 
     public static SdkAsyncHttpClient createAsyncHttpClient(
             final AttributeMap config, final NettyNioAsyncHttpClient.Builder httpClientBuilder) {
+        return createAsyncHttpClient(config, httpClientBuilder, SdkEventLoopGroup.builder());
```

Comment:
Please confirm that defaults are actually used here. I followed the code through and arrived here, when using a DefaultBuilder without setting any values. Seems to default to 0 threads:
- https://github.com/aws/aws-sdk-java-v2/blob/52a737b464d6d91247f1c53cfc994eeaeb375b5f/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/SdkEventLoopGroup.java#L127

Reply:
I think this should be lowered. This will basically mean that for N tasks (running in the same process), we'll have `N x 2 x num cpus` threads. It feels that some fairly low number should do for most use cases and we should allow overrides from the user.

WDYT?

Reply:
Right, I see. I elected to use the default as that's what the legacy `flink-connector-kinesis` (based on KPL) was using. Your thoughts, @dannycranmer ?

Reply:
The legacy connector based on KPL is not using async AWS SDK v2 client. Current (non FLIP-27) source is using async clients for EFO. This source creates [one client per sub task too](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java#L481), so this change will result in a larger number of threads being created and could cause a regression.

Since the Kinesis EFO source is currently using `2 x num cpus` I agree that `N x 2 x num cpus` is  big change.

> It feels that some fairly low number should do

I agree.  How about we pick 2 threads? This would follow the formula to 1 parallelism per CPU.

The other option is to try to calculate it dynamically to scale above 2 for bigger system, for instance:

```
thread_pool_size = max(2, (num_cpus / sink_parallelism) * 2)

# 2 CPUs at 8 parallelism (under provisioned)
thread_pool_size = max(2, (2 / 8) * 2) = 2

# 4 CPUs at 4 parallelism (target provisioning)
thread_pool_size = max(2, (4 / 4) * 2) = 2

# 32 CPUs at 4 parallelism (over provisioned)
thread_pool_size = max(2, (32 / 4) * 2) = 16
```

Maximum thread pool size here would be `2 x num cpus`. But does not scale for apps with many sinks, or clusters with many jobs.

> we should allow overrides from the user

I agree